rpc: Sanitize label name in various RPCs with tests

aureleoules opened the pull request:
The following RPCs did not sanitize the optional label name:
- importprivkey
- importaddress
- importpubkey
- importmulti
- importdescriptors
- listsinceblock

Thus is was possible to import an address with a label `*` which should not be possible.
The wildcard label is used for backwards compatibility in the `listtransactions` rpc.
I added test coverage for these RPCs.

aureleoules committed:
test: Invalid label name coverage

stickies-v reviewed (commented):
Approach ACK 3012200bc0602fc35a53b6a25a7d23eaaed54f9c

aureleoules commented:
Thanks for the review @stickies-v, I've addressed all your comments.

stickies-v commented on test/functional/wallet_labels.py:
```diff
@@ -25,6 +25,42 @@ def set_test_params(self):
+        for rpc_call in rpc_calls:
+            assert_raises_rpc_error(-11, "Invalid label name", *rpc_call, "*")
+
+        r = None
```
I don't think this line adds much, it's not the expected type of r and it's not necessary to declare it before the if/else. I'd prefer to remove, not really pythonic.

stickies-v commented on test/functional/wallet_labels.py:
```diff
@@ -25,6 +25,42 @@ def set_test_params(self):
+            }])
+        else:
+            privkey = node.dumpprivkey(address)
+            assert_raises_rpc_error(-11, "Invalid label name", node.importprivkey, privkey, "*")
```
nit: if you move the entire if/else loop to before the execution of the rpc calls, you could replace this with `rpc_calls.append(...)`

aureleoules commented:
Pushed, thanks @stickies-v. 
![](https://cdn.discordapp.com/emojis/966020215764844605.gif?size=4096)

stickies-v commented on test/functional/wallet_labels.py:
```diff
@@ -25,6 +25,41 @@ def set_test_params(self):
+            [node.getreceivedbylabel],
+        ]
+        if self.options.descriptors:
+            r = node.importdescriptors([{
```
nit: would avoid using one-letter variables, the brevity isn't necessary here imo and it's just slightly more confusing
```suggestion
            response = node.importdescriptors([{
```

stickies-v commented on test/functional/wallet_labels.py:
```diff
@@ -25,6 +25,41 @@ def set_test_params(self):
+        ]
+        if self.options.descriptors:
+            r = node.importdescriptors([{
+                'desc': 'pkh(' + address + ')',
```
nit: f-string
```suggestion
                'desc': f'pkh({address})',
```

stickies-v reviewed (approved):
ACK d29358558 - nice improvement, straightforward code changes and test.

Left a few nits, nothing blocking.

MarcoFalke commented on src/wallet/rpc/backup.cpp:
```diff
@@ -431,7 +431,7 @@ RPCHelpMan importpubkey()
 
     std::string strLabel;
     if (!request.params[1].isNull())
-        strLabel = request.params[1].get_str();
+        strLabel = LabelFromValue(request.params[1]);
```
What about moving the null check into `LabelFromValue` (if needed with an optional arg to set the default value to something other than `""`)?

aureleoules replied:
What do you think of 62535d877c0d9e092499b52c466b982b18b5df39?

maflcko reviewed (commented):
Concept ACK. Left a follow-up idea

jonatack commented on test/functional/wallet_labels.py:
```diff
@@ -25,6 +25,40 @@ def set_test_params(self):
+            [node.getaddressesbylabel],
+            [node.importpubkey, pubkey],
+            [node.addmultisigaddress, 1, [pubkey]],
+            [node.getreceivedbylabel],
```
Looks like listsinceblock ought to be added to this test if #25934 is merged.

aureleoules replied:
answered here: https://github.com/bitcoin/bitcoin/pull/25934#discussion_r984481844

stickies-v commented on src/wallet/rpc/coins.cpp:
```diff
@@ -21,7 +21,7 @@ static CAmount GetReceived(const CWallet& wallet, const UniValue& params, bool b
     if (by_label) {
         // Get the set of addresses assigned to label
-        addresses = wallet.ListAddrBookAddresses(CWallet::AddrBookFilter{LabelFromValue(params[0])});
+        addresses = wallet.ListAddrBookAddresses(CWallet::AddrBookFilter{CHECK_NONFATAL(LabelFromValue(params[0]))});
```
Since `AddrBookFilter` takes a `std::optional<std::string> m_op_label`, I think there's no need for the `CHECK_NONFATAL` here? Wouldn't be any behaviour change either I think, because the existing code would throw if params[0] were to be null? (not 100% sure on that)

```suggestion
        addresses = wallet.ListAddrBookAddresses(CWallet::AddrBookFilter{LabelFromValue(params[0])});
```

stickies-v commented on src/wallet/rpc/addresses.cpp:
```diff
@@ -140,7 +138,7 @@ RPCHelpMan setlabel()
     }
 
-    std::string label = LabelFromValue(request.params[1]);
+    auto label = *CHECK_NONFATAL(LabelFromValue(request.params[1]));
```
Missing `#include <util/check.h>` here and in other places.

stickies-v reviewed (approved):
ACK 8b716611b

MarcoFalke commented:
Removed from the milestone for now, as this is not a regression. Though, it can still be re-added and put into rc3 if merged in time.

theStack commented on test/functional/wallet_labels.py:
```diff
@@ -25,6 +25,40 @@ def set_test_params(self):
+                'timestamp': 'now',
+            }])
+        else:
+            rpc_calls.append([node.importprivkey, node.dumpprivkey(address), "*"])
```
Since the asterisk is added in the for-loop below, there is no need to add it here. I'd suggest to also add test coverage for `importaddress` here:
```suggestion
            rpc_calls.extend([
                [node.importprivkey, node.dumpprivkey(address)],
                [node.importaddress, address],
            ])
```

aureleoules replied:
Thanks done

theStack reviewed (commented):
Concept ACK

theStack reviewed (approved):
ACK 493cadc8ee0339244b722b2c36dac078a3397800

aureleoules commented:
Rebased to add coverage to `listsinceblock` and squashed 493cadc8ee0339244b722b2c36dac078a3397800 into the first commit.

theStack reviewed (approved):
re-ACK 21b429632d15a170317a73fb5b7aa3cf95347c78

furszy commented on src/wallet/wallet.h:
```diff
@@ -107,6 +107,7 @@ static const bool DEFAULT_WALLET_RBF = true;
 static const bool DEFAULT_WALLETBROADCAST = true;
 static const bool DEFAULT_DISABLE_WALLET = false;
 static const bool DEFAULT_WALLETCROSSCHAIN = false;
+static const std::string DEFAULT_WALLET_LABEL = "";
```
Conceptually there isn't a default label, the record label is just empty.

What if you return from `LabelFromValue` the empty string instead? It will also remove the need for all the `.value_or(DEFAULT_WALLET_LABEL);`.

aureleoules replied:
Some RPCs require the label to be optional for filtering. `getaddressesbylabel` will return all addresses if the label is nullopt for example.
I could hard-code the `""` in `.value_or` calls if you prefer though.

furszy replied:
Maybe you are referring to `listsinceblock`? (as far as I can see, it's the only place where the label is used as optional), `getaddressesbylabel` uses a string not an optional.

Check this out: https://github.com/furszy/bitcoin-core/commit/8ce2d8965537a889fe88ca1a891763b35aba32ad. (If you like it, feel free to squash it on the first commit).

aureleoules replied:
The CI doesn't pass on your commit because of `listsinceblock`. A label can be empty in this RPC.

https://cirrus-ci.com/task/6056876765347840?logs=ci#L3626

furszy replied:
will check it tonight 👍🏼. I coded a quick example. Shouldn't be hard to correct that single case.

furszy replied:
Ready, fix was simple: https://github.com/furszy/bitcoin-core/commit/f8da21cb6eec0d62b2d3ed766201aca182b47342.

stickies-v reviewed (approved):
re-ACK 21b429632

2 non-blocking nits

luke-jr commented:
You seem to have squashed the refactor into the fix commit. Please don't do that and re-separate them.

aureleoules commented:
Thanks @furszy I cherry-picked, squashed your commit, and added you as co-author.

> You seem to have squashed the refactor into the fix commit. Please don't do that and re-separate them.

The refactor commit is now gone thanks to furszy.

stickies-v commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -635,10 +635,9 @@ RPCHelpMan listsinceblock()
-        filter_label = request.params[5].get_str();
-    }
+    // Only set it if 'label' was provided.
+    std::optional<std::string> filter_label = request.params[5].isNull() ? std::nullopt :
+                                              std::optional(LabelFromValue(request.params[5]));
```
nit: could avoid multiline ternary this way:
```suggestion
    std::optional<std::string> filter_label;
    if (!request.params[5].isNull()) filter_label.emplace(LabelFromValue(request.params[5]));
```

stickies-v commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1163,7 +1158,7 @@ static UniValue ProcessImport(CWallet& wallet, const UniValue& data, const int64
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Internal addresses should not have a label");
         }
-        const std::string& label = data.exists("label") ? data["label"].get_str() : "";
+        std::string label{LabelFromValue(data["label"])};
```
This refactor introduces quite a few string copy operations, where previously references were passed. I think this can be quite easily avoided with the below patch. There are other callsites of `LabelFromValue` that would benefit from just changing their type to `const std::string&`, but to keep LoC change minimal I've not changed them in this patch.

<details>
<summary>git diff on ba1e78a2d58059f439f17e821eb2d1bab4db86cb</summary>

```diff
diff --git a/src/wallet/rpc/addresses.cpp b/src/wallet/rpc/addresses.cpp
index e819c09c5..5d04d3113 100644
--- a/src/wallet/rpc/addresses.cpp
+++ b/src/wallet/rpc/addresses.cpp
@@ -43,7 +43,7 @@ RPCHelpMan getnewaddress()
     }
 
     // Parse the label first so we don't generate a key if there's an error
-    std::string label{LabelFromValue(request.params[0])};
+    const std::string& label{LabelFromValue(request.params[0])};
 
     OutputType output_type = pwallet->m_default_address_type;
     if (!request.params[1].isNull()) {
@@ -256,7 +256,7 @@ RPCHelpMan addmultisigaddress()
 
     LOCK2(pwallet->cs_wallet, spk_man.cs_KeyStore);
 
-    std::string label{LabelFromValue(request.params[2])};
+    const std::string& label{LabelFromValue(request.params[2])};
 
     int required = request.params[0].getInt<int>();
 
diff --git a/src/wallet/rpc/backup.cpp b/src/wallet/rpc/backup.cpp
index 1abb2efda..7e5fbfae0 100644
--- a/src/wallet/rpc/backup.cpp
+++ b/src/wallet/rpc/backup.cpp
@@ -140,7 +140,7 @@ RPCHelpMan importprivkey()
         EnsureWalletIsUnlocked(*pwallet);
 
         std::string strSecret = request.params[0].get_str();
-        std::string strLabel{LabelFromValue(request.params[1])};
+        const std::string& strLabel{LabelFromValue(request.params[1])};
 
         // Whether to perform rescan after import
         if (!request.params[2].isNull())
@@ -232,7 +232,7 @@ RPCHelpMan importaddress()
     EnsureLegacyScriptPubKeyMan(*pwallet, true);
 
     // Parse label
-    std::string strLabel{LabelFromValue(request.params[1])};
+    const std::string& strLabel{LabelFromValue(request.params[1])};
 
     // Whether to perform rescan after import
     bool fRescan = true;
@@ -423,7 +423,7 @@ RPCHelpMan importpubkey()
 
     EnsureLegacyScriptPubKeyMan(*pwallet, true);
 
-    std::string strLabel{LabelFromValue(request.params[1])};
+    const std::string& strLabel{LabelFromValue(request.params[1])};
 
     // Whether to perform rescan after import
     bool fRescan = true;
@@ -1158,7 +1158,7 @@ static UniValue ProcessImport(CWallet& wallet, const UniValue& data, const int64
         if (internal && data.exists("label")) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Internal addresses should not have a label");
         }
-        std::string label{LabelFromValue(data["label"])};
+        const std::string& label{LabelFromValue(data["label"])};
         const bool add_keypool = data.exists("keypool") ? data["keypool"].get_bool() : false;
 
         // Add to keypool only works with privkeys disabled
@@ -1452,7 +1452,7 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
         const std::string& descriptor = data["desc"].get_str();
         const bool active = data.exists("active") ? data["active"].get_bool() : false;
         const bool internal = data.exists("internal") ? data["internal"].get_bool() : false;
-        std::string label{LabelFromValue(data["label"])};
+        const std::string& label{LabelFromValue(data["label"])};
 
         // Parse descriptor string
         FlatSigningProvider keys;
diff --git a/src/wallet/rpc/transactions.cpp b/src/wallet/rpc/transactions.cpp
index 7088ff4ab..9b03782d4 100644
--- a/src/wallet/rpc/transactions.cpp
+++ b/src/wallet/rpc/transactions.cpp
@@ -636,8 +636,8 @@ RPCHelpMan listsinceblock()
     bool include_change = (!request.params[4].isNull() && request.params[4].get_bool());
 
     // Only set it if 'label' was provided.
-    std::optional<std::string> filter_label = request.params[5].isNull() ? std::nullopt :
-                                              std::optional(LabelFromValue(request.params[5]));
+    const std::optional<std::string>& filter_label = request.params[5].isNull() ? std::nullopt :
+                                                     std::optional{LabelFromValue(request.params[5])};
 
     int depth = height ? wallet.GetLastBlockHeight() + 1 - *height : -1;
 
diff --git a/src/wallet/rpc/util.cpp b/src/wallet/rpc/util.cpp
index 1fdb80630..bc0c284dc 100644
--- a/src/wallet/rpc/util.cpp
+++ b/src/wallet/rpc/util.cpp
@@ -130,11 +130,12 @@ const LegacyScriptPubKeyMan& EnsureConstLegacyScriptPubKeyMan(const CWallet& wal
     return *spk_man;
 }
 
-std::string LabelFromValue(const UniValue& value)
+const std::string& LabelFromValue(const UniValue& value)
 {
-    if (value.isNull()) return ""; // empty
+    static std::string empty_string;
+    if (value.isNull()) return empty_string;
 
-    std::string label = value.get_str();
+    const std::string& label{value.get_str()};
     if (label == "*")
         throw JSONRPCError(RPC_WALLET_INVALID_LABEL_NAME, "Invalid label name");
     return label;
diff --git a/src/wallet/rpc/util.h b/src/wallet/rpc/util.h
index 87d34f7c1..9dd101167 100644
--- a/src/wallet/rpc/util.h
+++ b/src/wallet/rpc/util.h
@@ -5,6 +5,7 @@
 #ifndef BITCOIN_WALLET_RPC_UTIL_H
 #define BITCOIN_WALLET_RPC_UTIL_H
 
+#include <attributes.h>
 #include <script/script.h>
 
 #include <any>
@@ -40,7 +41,7 @@ const LegacyScriptPubKeyMan& EnsureConstLegacyScriptPubKeyMan(const CWallet& wal
 
 bool GetAvoidReuseFlag(const CWallet& wallet, const UniValue& param);
 bool ParseIncludeWatchonly(const UniValue& include_watchonly, const CWallet& wallet);
-std::string LabelFromValue(const UniValue& value);
+const std::string& LabelFromValue(const UniValue& value LIFETIMEBOUND);
 //! Fetch parent descriptors of this scriptPubKey.
 void PushParentDescriptors(const CWallet& wallet, const CScript& script_pubkey, UniValue& entry);
 ```
</details>

luke-jr commented:
>The refactor commit is now gone thanks to furszy.

Please restore it.

aureleoules committed:
rpc: Sanitize label name in various RPCs

- importprivkey
- importaddress
- importpubkey
- listtransactions
- listsinceblock
- importmulti
- importdescriptors

aureleoules committed:
refactor: Add sanity checks in LabelFromValue

aureleoules commented:
> Please restore it.

Done @luke-jr.

furszy commented on src/wallet/rpc/util.h:
```diff
@@ -40,7 +41,7 @@ const LegacyScriptPubKeyMan& EnsureConstLegacyScriptPubKeyMan(const CWallet& wal
 bool GetAvoidReuseFlag(const CWallet& wallet, const UniValue& param);
 bool ParseIncludeWatchonly(const UniValue& include_watchonly, const CWallet& wallet);
-std::string LabelFromValue(const UniValue& value);
+const std::string& LabelFromValue(const UniValue& value LIFETIMEBOUND);
```
Guess that you wanted to set `LIFETIMEBOUND` for the return type instead of the function's argument right?.

Still, would return the plain string and not a reference. There is no need to overcomplicate code for address book labels.

aureleoules replied:
Thanks, removed the reference.

stickies-v replied:
> Guess that you wanted to set LIFETIMEBOUND for the return type instead of the function's argument right?

The return value of `LabelFromValue` was bound to the lifetime of `const UniValue& value` so the attribute was correct I think? See https://clang.llvm.org/docs/AttributeReference.html#lifetimebound

> There is no need to overcomplicate code for address book labels.

I'm honestly not sure why this got reverted. It introduces almost no additional complexity in the code, and callsites can now _choose_ if they wish to have `LabelFromValue` return by value or by reference. Even though wallet labels aren't expected to be huge strings, I don't see the downside of doing it this way? There's no point bikeshedding over this, we won't be able to measure the performance impact, but it just seems like the more proper way of doing things, imo. Up to you, @aureleoules.

furszy replied:
> I'm honestly not sure why this got reverted. It introduces almost no additional complexity in the code, and callsites can now choose if they wish to have LabelFromValue return by value or by reference. Even though wallet labels aren't expected to be huge strings, I don't see the downside of doing it this way? There's no point bikeshedding over this, we won't be able to measure the performance impact, but it just seems like the more proper way of doing things, imo. Up to you, @aureleoules.

I'm actually not really happy with the revert neither but for another reason.

But first, my point above was to take advantage of the compiler's [rvo](https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization), which is designed for this kind of stuff.

Now, the reason why I'm not particularly happy with the latest changes is the use of the `auto` keyword alone (which involves a string copy). Instead, should receive the function's output into a const ref string. @aureleoules.

aureleoules replied:
Is this ok with you? https://github.com/bitcoin/bitcoin/compare/cce2e872e1a8b7c5b80036f48f1f35467b9328ff..b119bb84250f5b97053314f730aefb8c10b4867a

furszy replied:
Have spoken with @stickies-v further and he made me notice that the string is not initialized inside `LabelFromValue`, the univalue `get_str` method returns a const reference.. so the compiler's rvo argument does not apply.

So all good for me now, I'm aligned on returning the const ref string from `LabelFromValue` and adding the lifetime bound attribute once more.
just need to re-add the following line:
```c++
const std::string& LabelFromValue(const UniValue& value LIFETIMEBOUND);
```

stickies-v replied:
Can be marked as resolved, as furszy mentioned we're all in agreement here and I'm okay with the current implementation 👍

ajtowns replied:
Changing this to return a reference doesn't seem very safe -- in that case calling `string& label = LabelFromValue(data["label"])` will mean `label` can become invalid if `data` gets changed (eg, if additional calls to `data.pushKV()` are made, the `data.values` vector may be reallocated, invalidating the reference), which is not something `LIFETIMEBOUND` can check.

As far as I can see we're talking one copy of each fairly short label per RPC call here, which isn't something that needs to be aggressively optimised. So seems much better to me to just keep the implicit copy of the string and have a simple API that minimises the risk of errors creeping in.

stickies-v replied:
That's a great point I hadn't considered. Even though the `data["label"]` `UniValue::VSTR` object that we bind to doesn't use a vector to store its string value and would seem safe at first, `data["label"]` returns a reference to the `data.values` vector - so any changes to that parent `UniValue::VOBJ` that we can't control for here are potentially dangerous.

I agree that this could lead to nasty issues and isn't worth the overhead. Sorry for causing the back and forth on this @aureleoules. The `UniValue` type can be a bit tricky to reason about, it seems.

aureleoules replied:
Done https://github.com/bitcoin/bitcoin/compare/408e375d15e327b2f5886f395dbc23a4834460e5..632202c5b7bb5fb975f8bde74b6085772b142c5b

1440000bytes commented:
Concept ACK

stickies-v commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -487,7 +487,7 @@ RPCHelpMan listtransactions()
     std::optional<std::string> filter_label;
     if (!request.params[0].isNull() && request.params[0].get_str() != "*") {
-        filter_label = request.params[0].get_str();
+        filter_label = LabelFromValue(request.params[0]);
```
nit
```suggestion
        filter_label.emplace(LabelFromValue(request.params[0]));
```

aureleoules replied:
Done

stickies-v reviewed (approved):
re-ACK 408e375d1

Reduced string copying compared to my previous ACK. One non-blocking nit.

furszy reviewed (approved):
ACK 893c2888

ajtowns reviewed (commented):
Concept ACK.

furszy commented on src/wallet/rpc/util.h:
```diff
@@ -5,6 +5,7 @@
 #ifndef BITCOIN_WALLET_RPC_UTIL_H
 #define BITCOIN_WALLET_RPC_UTIL_H
 
+#include <attributes.h>
```
In 7d2d0c4d:
is this include intended?

aureleoules replied:
Removed thanks

furszy reviewed (approved):
diff ACK 65e78bd

stickies-v commented on src/wallet/rpc/util.cpp:
```diff
@@ -132,7 +132,10 @@ const LegacyScriptPubKeyMan& EnsureConstLegacyScriptPubKeyMan(const CWallet& wal
 std::string LabelFromValue(const UniValue& value)
 {
-    std::string label = value.get_str();
+    static const std::string empty_string;
+    if (value.isNull()) return empty_string;
```
nit: not needed anymore now that we're returning a string copy

```suggestion
    if (value.isNull()) return "";
```

ajtowns replied:
Or `return {};`

stickies-v reviewed (approved):
re-ACK 65e78bda7cd23ad7b6ede63c6bf5ffe6f552c71d

ajtowns commented:
ACK 65e78bda7cd23ad7b6ede63c6bf5ffe6f552c71d

theStack reviewed (approved):
re-ACK 65e78bda7cd23ad7b6ede63c6bf5ffe6f552c71d

achow101 commented on test/functional/wallet_labels.py:
```diff
@@ -28,6 +28,44 @@ def set_test_params(self):
+            [node.getnewaddress],
+            [node.setlabel, address],
+            [node.getaddressesbylabel],
+            [node.importpubkey, pubkey],
```
In 65e78bda7cd23ad7b6ede63c6bf5ffe6f552c71d "test: Invalid label name coverage"

`importpubkey` is a legacy wallet only RPC. The reason this test still passes with `--descriptors` is because the test framework has this (and some of the other import RPCs) aliased to `importdescriptors`.

aureleoules replied:
Considering this PR has 4 ACKs is it worth the change? The test is executed with both `--legacy-wallet` and `--descriptors` anyway.

achow101 commented:
ACK 65e78bda7cd23ad7b6ede63c6bf5ffe6f552c71d
